### ISSUE TITLE
Enumerator::Lazy#sum and Enumerator#sum behave differently after Rails 7 update if block relies on side effect

### ISSUE THREAD

**Issue by goalaleo** [rails/rails#48712]
### Steps to reproduce
I'm doing an update from rails `6.1.7.2` to `7.0.6` and some of our tests failed. This seems related to the deprecation of `Enumerable#sum` in Rails 7.

Basically the problem is that `Enumerator::Lazy#sum` and `Enumerator#sum` - when used without initial value - behave differently if the block relies on a side effect.
```ruby
input1 = "\xEF\xBB\xBFinput1"
input2 = "\xEF\xBB\xBFinput2"

# returns a lazy enumerator
remove_first_bom = proc do |enum|
  # external state
  first = true
  
  enum.lazy.map do |str|
    if first
      # side effect
      first = false
      str[1..]
    else
      str
    end
  end
end

# when using deprecated Enumerator::Lazy#sum the BOM bytes [239, 187, 191] are not removed.
remove_first_bom.call([input1, input2]).sum.bytes
=> [239, 187, 191, 105, 110, 112, 117, 116, 49, 239, 187, 191, 105, 110, 112, 117, 116, 50]

# when using deprecated Enumerator#sum the BOM bytes are removed as expected when
# lazy enumerator is forced to evaluate before using sum
remove_first_bom.call([input1, input2]).force.sum.bytes
=> [105, 110, 112, 117, 116, 49, 239, 187, 191, 105, 110, 112, 117, 116, 50]

# when the initial value for sum is provided this seems to work, so in that sense not using the deprecated behaviour fixes the issue
remove_first_bom.call([input1, input2]).sum("").bytes
 => [105, 110, 112, 117, 116, 49, 239, 187, 191, 105, 110, 112, 117, 116, 50] 
```
I suspect that there might be some peeking behaviour which has the side effect of setting `first = false`. I'm pretty sure I saw this when adding `puts` into the conditional.

I confirmed that this doesn't happen in `6.1.7.2`.

### Expected behavior
`Enumerator::Lazy#sum` and `Enumerator#sum` should return the same result.

### Actual behavior
`Enumerator::Lazy#sum` and `Enumerator#sum` return different values if the block relies on a side effect

### System configuration
**Rails version**: `7.0.6`

**Ruby version**:  `3.1.2`

**Comment by goalaleo** [comment]
@fatkodima here's a repro script
```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", "7.0.6"
end

require "active_support/core_ext/enumerable"
require "minitest/autorun"

class BugTest < Minitest::Test
  def test_enumerable_lazy_sum_with_side_effect
    input1 = "\xEF\xBB\xBFinput1"
    input2 = "\xEF\xBB\xBFinput2"

    # Remove BOM from str if it's the first element in the enumerable.
    # Subsequent elements are not modified. Relies on the side effect of
    # changing the variable `first` to false.
    remove_first_bom = proc do |enum|
      first = true
      enum.lazy.map do |str|
        if first
          first = false
          str[1..]
        else
          str
        end
      end
    end


    # using a lazy enumerator, this fails
    assert_equal "input1" + input2, remove_first_bom.call([input1, input2]).sum
    # forcing the enumerator to be evaluated before sum, this works
    assert_equal "input1" + input2, remove_first_bom.call([input1, input2]).force.sum
  end
end
```

**Comment by fatkodima** [comment]
`git bisect` shows that the change was introduced in https://github.com/rails/rails/pull/42080. As the warning from the bug report execution shows, you need to pass an argument to `.sum` (like `.sum('')`) and the test then passes.

This monkey patching of `Enumerable` and `Array` is already removed from the `main` (which will be included in the `7.1` release soon), so I am not sure we should do anything about it.

**Comment by goalaleo** [comment]
Yep, I also noticed that calling Ruby stdlib's `sum('')` with the initial argument makes the test pass.

Strictly speaking deprecated behaviour should still work as before until it's fully removed - at least this is how I understand it. Whether it's worth it to fix this is ofc up to the maintainers, but I wanted to open the issue in case someone else faces the same problem. We've fixed the issue on our end by not relying on a side effect in the block.
